packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts

```diff
@@ -85,39 +94,66 @@ export const useGridDataSource = (
         apiRef.current.resetDataSourceState();
       }
 
-      const fetchParams = gridGetRowsParamsSelector(apiRef);
-
+      const fetchParams = { ...gridGetRowsParamsSelector(apiRef), ...rowFetchSlice.current };
+      const startingIndex = fetchParams.start;
       const cachedData = apiRef.current.unstable_dataSource.cache.get(fetchParams);
 
       if (cachedData !== undefined) {
         const rows = cachedData.rows;
-        apiRef.current.setRows(rows);
-        if (cachedData.rowCount) {
+        if (cachedData.rowCount !== undefined) {
           apiRef.current.setRowCount(cachedData.rowCount);
         }
+        if (isLazyLoaded) {
+          apiRef.current.unstable_replaceRows(startingIndex, rows);
+        } else {
+          apiRef.current.setRows(rows);
+        }
+        apiRef.current.publishEvent('rowsFetched');
         return;
       }
 
+      // with lazy loading, only the initial load should show the loading overlay
+      const useLoadingIndicator = !isLazyLoaded || apiRef.current.getRowsCount() === 0;
       const isLoading = gridRowsLoadingSelector(apiRef);
-      if (!isLoading) {
+      if (!isLoading && useLoadingIndicator) {
         apiRef.current.setLoading(true);
       }
 
       try {
         const getRowsResponse = await getRows(fetchParams);
         apiRef.current.unstable_dataSource.cache.set(fetchParams, getRowsResponse);
-        if (getRowsResponse.rowCount) {
+        if (getRowsResponse.rowCount !== undefined) {
```

Comment:
_As discussed already, documenting here for further discussion:_

I think the only difference as an end user between **lazy loading** and **infinite loading** would be the presence of `rowCount`.

For example, in an implementation, the `rowCount` is not available in the beginning, or it’s an unknown row count (-1), the Data Grid works in an infinite loading mode, but at a later point, the `rowCount` becomes available. At this point, the Data Grid has to compute the total viewport height and shift to a `lazyLoading` mode. This can also happen the other way around.

Could we remove the line between both cases based on this assumption, and handle them together?

Reply:
> we will definitely have to add an explanation in the docs that one server side data feature replaces two client site features

The idea was to deprecate the existing features with the introduction of the data source variants and remove them eventually. Because it doesn't make much sense to have these features used on the client side.

I'm confident that even today, most of the implementations on the users' side on top of the current lazy loading and infinite loading are binded to server-side logic on the user's side.

Even if there is some use-case on the client-side (for example the users want to lazy load a large JSON file present locally), it should be possible by using the Data source interface (implementing the `getRows` method on the client-side).

What do you think?

> let's say that you load initial page and scroll fast to 3rd page or more. This will leave some skeleton rows in between the content, so switching to `infiniteLoading` will not make it possible to load those rows anymore (with the current way of working)

For `lazy-loading` => `infinite loading` transition (setting of `rowCount` prop to `undefined` or `-1`) I'd consider the existing rows **stale**. I'd reset the rows, move the user to the top, and send the request for the first page since we are unsure about the row count at this point, it might be less than already loaded lazy-loaded rows.

While a transition `infinite loading` => `lazy-loading` (setting of `rowCount` prop from `undefined` or `-1` to some finite number) might require some further processing based on the provided `rowCount` value. 

There could be 3 possibilities in this regard. (consider _pageSize/batchSize = 10_ in the examples)

**rowCount > loaded rows**
Let's say the loaded rows are **1-60** and the `rowCount` is provided **100**, we just need to add 40 skeleton rows below the current rows and keep the existing rows rendered.

**rowCount == loaded rows**
If 60 rows are loaded and the rowCount is set to 60, we remove the 10 skeleton rows (rendered due to infinite loading) and consider all the data to be fetched.

**rowCount < loaded rows**
I'm doubtful about this case. It could mean one of the following:
1. _The data set is changed:_ In this case, we need to reset the rows, move the user to the top, and send the request for the first page again.
2. _The data set is modified, and some records are deleted from the end:_ We can simply discard the extra rows) in this case (for example, 1-60 are loaded, `rowCount` provided is 50, we internally delete rows 51-60, move the scroll to page 5 (let's say row 41) and consider all the rows to be fetched at this point
3. _The data set is modified, and some records are deleted from arbitrary places:_ We'll have to reset the rows and re-fetch the data from the beginning again)

I'm not sure how we differentiate between these cases and handle them (in terms of the public API)

Reply:
I have added the support for infinite loading as discussed.
if `rowCount` is `undefined` or `-1`, infinite loading is in place
otherwise, lazy loading kicks in.

Since we want to have them combined I thought of using the same name (lazy loading) for both and differentiate them as different modes of lazy loading (viewport and infinite)
WDYT?

I still have to check the scenario where the modes are changed after initial load

Reply:
Changing loading mode is now supported (also via api and via the prop)
Logic to determine when the grid should be reset is explained in the code comments

https://github.com/mui/mui-x/blob/cc4423d8bfa824cc92a10088efadbb152292a422/packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridDataSourceLazyLoader.ts#L81-L108

it is aligned with what you have suggested as well
lazy loading -> infinite loading resets the grid
infinite loading -> lazy loading
- adds skeleton rows to fill the gap (in case the count is the same nothing happens)
- if the count is less than the amount of rows, gird is reset

I will add an example in the docs that will be more of a demonstration of this behavior, but there will be a note that it is not expected that you constantly update the row count, especially that it does not get reduced to a number smaller than the amount of the rows already rendered

Reply:
In addition to this, grid resets even if you stay on lazy loading mode but you render certain amount of rows and then update the `rowCount` prop to be lower than that